ENH - New ``skglm`` documentation (#118)
scikit-learn-contrib/skglm: doc/getting_started.rst

.. _getting_started:

===============
Getting started
===============
---------------

This page provides a started examples to get familiar with ``skglm``
and explore some of its features.

We start by fitting a Lasso estimator on a high dimensional (number of features is largely greater than the number of samples)

> toy dataset using ``skglm`` Lasso estimator. and then we explore how modularity of the package by building 

Reviewer: ```suggestion
toy dataset using ``skglm`` Lasso estimator, and then we explore how modularity of the package by building 
```

Beforehand, make sure that you have already installed ``skglm`` either through

.. code-block:: shell

    # using pip
    pip install -U skglm

    # using conda
    conda install skglm

-------------------------

Fitting a Lasso estimator
-------------------------

Let's start first by generating a toy dataset and splitting it to train and test sets.
For that, we will use ``scikit-learn`` 
`make_regression <https://scikit-learn.org/stable/modules/generated/sklearn.datasets.make_regression.html#sklearn.datasets.make_regression>`_

.. code-block:: python

    # imports
    from sklearn.datasets import make_regression
    from sklearn.model_selection import train_test_split

    # generate toy data
    X, y = make_regression(n_samples=100, n_features=1000)
    
    # split data
    X_train, X_test, y_train, y_test = train_test_split(X, y)

Then let's fit ``skglm`` :ref:`Lasso <skglm.Lasso>` estimator and prints it's score on the test set.

.. code-block:: python

    # import estimator
    from skglm import Lasso
    
    # init and fit
    estimator = Lasso()
    estimator.fit(X_train, y_train)

    # compute R²
    estimator.score(X_test, y_test)

.. note::

>     - Every first fit of ``skglm`` estimator take some extra time as ``skglm`` uses `Numba <https://numba.pydata.org/>`_ 

Reviewer: W need to be careful here because it our major weakness at the moment. We should make it clear that this overhead is only for the first fit after importing ``skglm``. If you import Lasso from skglm and git (you have an overhead), if you then import GLE with Huber and L1 and fit the estimator, you don't. So we should make it as clear as possible, not to shoot ourselves in the foot.

      to just-in-time compile the code.
      Afterward, the other fits will be super fast.

``skglm`` has several other ``scikit-learn`` compatible estimators.
Check the :ref:`API <Estimators>` for more information about the available estimators.

Fitting :math:`\ell_1` Huber regression
---------------------------------------

Suppose that we have a dataset with outliers and we would like to mitigate their effects on the learned coefficients
while considering the high dimensionality of dataset. Ideally, we would like to fit a :math:`\ell_1` Huber regressor.

> Unfortunately, this estimator is missing from ``scikit-learn`` and ``skglm`` already available estimators.

Author: Can't disagree!

``skglm`` offers high flexibility to compose custom estimators. Through a simple API, it is possible to combine any
``skglm`` :ref:`datafit <Datafits>` and :ref:`penalty <Penalties>`.

Let's explore how to achieve that.

Generate corrupted data
***********************

We will use the same script as before except that we will take 10 samples and corrupt their values.

.. code-block:: python

    # imports
    import numpy as np
    from sklearn.datasets import make_regression
    from sklearn.model_selection import train_test_split

    # generate toy data
    X, y = make_regression(n_samples=100, n_features=1000)

    # select and corrupt 10 random samples
    y[np.random.choice(n_samples, 10)] = 100 * y.max()

    # split data
    X_train, X_test, y_train, y_test = train_test_split(X, y)

> Now let's compose a custom estimator using :ref:`GeneralizedLinearEstimator <skglm.GeneralizedLinearEstimator>`.

Reviewer: I'd add one or two sentences on what is GLE, and why it is important. Something like "it is the go-to way when one wants to create custom estimator by combining a datafit and a penalty".

.. code-block:: python

    # import penalty and datafit
    from skglm.penalties import L1
    from skglm.datafits import Huber

    # import GLM estimator
    from skglm import GeneralizedLinearEstimator

    # build and fit estimator
    estimator = GeneralizedLinearEstimator(
        Huber(1.),
        L1(alpha=1.)
    )
    estimator.fit(X_train, y_train)

.. note::

    Here the arguments given to the datafit and penalty are arbitrary and given just for sake of illustration.

> It is possible to combine any supported datafit and penalty. Explore the list of supported :ref:`datafits <Datafits>` 

Reviewer: ```suggestion
``GeneralizedLinearEstimator`` allows to combine any penalties and datafits implemented in ``skglm``. If you don't find an estimator in the ``estimators`` module, you can build it by combining the appropriate datafit and penalty and pass it to ``GeneralizedLinearEstimator``. Explore the list of supported :ref:`datafits <Datafits>` 
```

and :ref:`penalties <Penalties>`.

.. important::

>     It is possible to create custom datafit and penalties.

Reviewer: ```suggestion
    It is possible to create your own datafits and penalties.
```

    Check the tutorials on :ref:`how to add a custom datafit <how_to_add_custom_datafit>` and
    :ref:`how to add a custom penalty <how_to_add_custom_penalty>`.

> Explore further advanced topics and get hand-on examples on the :ref:`tutorials page <tutorials>`

Reviewer: ```suggestion
Explore further advanced topics and get hands-on examples on the :ref:`tutorials page <tutorials>`
```

.. _getting_started:

===============
Getting started
===============
---------------

This page provides a starter example to get familiar with ``skglm`` and explore some of its features.

In the first section, we fit a Lasso estimator on a high dimensional

> toy dataset (number of features is largely greater than the number of samples). Linear models doesn't generalize well

Reviewer: ```suggestion
toy dataset (number of features is largely greater than the number of samples). Linear models don't generalize well
```
